rtmr/extend.go

```diff
@@ -21,22 +21,55 @@ import (
 	"crypto"
 	_ "crypto/sha512" // Register SHA384 and SHA512
 	"fmt"
+	"os"
 
 	"github.com/google/go-configfs-tsm/configfs/configfsi"
 	"github.com/google/go-configfs-tsm/configfs/linuxtsm"
 	"github.com/google/go-configfs-tsm/rtmr"
 )
 
-// ExtendDigestClient extends the measurement to the rtmr with the given client.
+// ExtendDigestSysfs extends the measurement to the RTMR through the sysfs interface.
+// This is the modern method for extending TDX RTMRs.
+func ExtendDigestSysfs(rtmrIndex int, digest []byte) error {
+	// The sysfs file is an interface for RTMR. Each write operation functions
+	// as an "extend" operation, not a normal file write.
+	filePath := fmt.Sprintf("/sys/class/misc/tdx_guest/measurements/rtmr%d:sha384", rtmrIndex)
+	file, err := os.OpenFile(filePath, os.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("failed to open TDX RTMR file %s: %w", filePath, err)
+	}
+	defer file.Close()
+
+	if _, err = file.WriteAt(digest, 0); err != nil {
+		return fmt.Errorf("failed to write data to %s at offset 0: %w", filePath, err)
+	}
+	return nil
+}
+
+// ExtendDigest dynamically chooses the write method to extend the measurement to the RTMR
+// based on the system environment. It will try the legacy configfs path first and
+// fall back to the sysfs method if it doesn't exist.
 func ExtendDigestClient(client configfsi.Client, rtmrIndex int, digest []byte) error {
 	if rtmrIndex < 0 || rtmrIndex > 3 {
 		return fmt.Errorf("invalid rtmr index %d. For TDX, index can only be 0-3", rtmrIndex)
 	}
+
 	if len(digest) != crypto.SHA384.Size() {
 		return fmt.Errorf("sha384 digest should be %d bytes, the input is %d bytes", crypto.SHA384.Size(), len(digest))
 	}
-	// TODO: check the TCG mapping of the rtmr index
-	return rtmr.ExtendDigest(client, rtmrIndex, digest)
+
+	const legacyRTMRPath = "/sys/kernel/config/tsm/rtmrs"
+	// Check if the legacy configfs path exists to determine the extension method.
+	if _, err := os.Stat(legacyRTMRPath); err == nil {
+		// TODO: check the TCG mapping of the rtmr index
+		return rtmr.ExtendDigest(client, rtmrIndex, digest)
+	} else if !os.IsNotExist(err) {
+		// Handle unexpected errors from os.Stat, other than the path not existing.
+		return fmt.Errorf("failed to check for directory %s: %w", legacyRTMRPath, err)
+	}
+
+	// If the configfs path does not exist, use the sysfs interface instead.
+	return ExtendDigestSysfs(rtmrIndex, digest)
```

Comment:
The logic that picks sysfs/configfs can be in the ExtendDigest()(at the end of this file). We should probably keep the ExtendDigestClient function intact